examples/source_separation/lightning_trainer.py

```diff
@@ -0,0 +1,432 @@
+        lr_scheduler: The LR Scheduler, optional.
+    """
+
+    def __init__(
+        self,
+        model: Any,
+        train_loader: DataLoader,
+        val_loader: Optional[DataLoader],
+        loss: Any,
+        optim: Any,
+        metrics: List[Any],
+        lr_scheduler: Optional[Any] = None,
+    ) -> None:
+        super().__init__()
+
+        self.model: nn.Module = model
+        self.loss: nn.Module = loss
+        self.optim: torch.optim.Optimizer = optim
+        self.lr_scheduler: Optional[_LRScheduler] = None
+        if lr_scheduler:
+            self.lr_scheduler = lr_scheduler
+
+        self.metrics: Mapping[str, nn.Module] = metrics
+
+        self.train_metrics: nn.ModuleDict = nn.ModuleDict()
+        self.val_metrics: nn.ModuleDict = nn.ModuleDict()
+        self.test_metrics: nn.ModuleDict = nn.ModuleDict()
+
+        self.save_hyperparameters()
+        self.train_loader = train_loader
+        self.val_loader = val_loader
+
+    def setup(self, stage: Optional[str] = None) -> None:
+        if stage == "fit":
+            self.train_metrics.update(self.metrics)
+            self.val_metrics.update(self.metrics)
+        else:
+            self.test_metrics.update(self.metrics)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Forward defines the prediction/inference actions.
+        """
+        return self.model(x)
+
+    def training_step(
+        self, batch: Batch, batch_idx: int, *args: Any, **kwargs: Any
+    ) -> Dict[str, Any]:
+        return self._step(batch, batch_idx, "train")
+
+    def validation_step(
+        self, batch: Batch, batch_idx: int, *args: Any, **kwargs: Any
+    ) -> Dict[str, Any]:
+        """
+        Operates on a single batch of data from the validation set.
+        """
+        return self._step(batch, batch_idx, "val")
+
+    def test_step(
+        self, batch: Batch, batch_idx: int, *args: Any, **kwargs: Any
+    ) -> Optional[Dict[str, Any]]:
+        """
+        Operates on a single batch of data from the test set.
+        """
+        return self._step(batch, batch_idx, "test")
+
+    def _step(self, batch: Batch, batch_idx: int, phase_type: str) -> Dict[str, Any]:
+        mix, src, mask = batch
+        pred = self.model(mix)
+        loss = self.loss(pred, src, mask)
+        self.log(f"Losses/{phase_type}_loss", loss.item(), on_step=True, on_epoch=True)
+
+        metrics_result = self._compute_metrics(pred, src, mix, mask, phase_type)
+        self.log_dict(metrics_result, on_epoch=True)
+
+        return loss
+
+    def configure_optimizers(
+        self,
+    ) -> Tuple[Any]:
+        lr_scheduler = self.lr_scheduler
+        if not lr_scheduler:
+            return self.optim
+        epoch_schedulers = {
+            'scheduler': lr_scheduler,
+            'monitor': 'Losses/val_loss',
+            'interval': 'epoch'
+        }
+        return [self.optim], [epoch_schedulers]
```

Comment:
I don't know PyTorchLightning well, but one thing we need to double-check is if `LightningModule` can auto-wrap our given `model` into other abstractions internally (like `DistributedDataParallel`) before calling `configure_optimizers`. If that's the case, then the optimizers / schedulers might point to the unwrapped model, leading to unexpected behavior